libdrgn/iterators.c

```diff
@@ -0,0 +1,472 @@
+	}
+	frames_deinit(&iter->frames);
+}
+
+struct drgn_error *radix_tree_iter_next(struct radix_tree_iter *iter,
+					struct radix_tree_iter_entry **ret)
+{
+	bool entry_populated = false;
+	bool node_is_truthy = false;
+	struct drgn_error *err = NULL;
+	struct drgn_object node;
+	drgn_object_init(&node, iter->prog);
+	if (!iter->started) {
+		iter->started = true;
+		err = radix_tree_iter_handle_node(iter, &iter->root, 0,
+						  &entry_populated);
+		if (!err && entry_populated)
+			err = drgn_object_bool(&iter->entry.node,
+					       &node_is_truthy);
+	}
+
+	while (!err && !entry_populated && iter->frames.size) {
+		struct radix_tree_iter_frame *frame =
+			&iter->frames.data[iter->frames.size - 1];
+		if (frame->loop_index <= iter->RADIX_TREE_MAP_MASK) {
+			err = drgn_object_subscript(&node, &frame->slots,
+						    frame->loop_index);
+			if (!err)
+				err = radix_tree_iter_handle_node(
+					iter, &node,
+					frame->index + (frame->loop_index++
+							<< frame->shift),
+					&entry_populated);
+		} else {
+			drgn_object_deinit(&frame->slots);
+			frames_pop(&iter->frames);
+		}
+	}
+	*ret = err || !entry_populated ? NULL : &iter->entry;
+	drgn_object_deinit(&node);
+	return err;
+}
+
+struct drgn_error *idr_iter_init(struct idr_iter *iter,
+				 const struct drgn_object *idr)
+{
+	struct drgn_object idr_rt;
+	struct drgn_object idr_base;
+	struct drgn_error *err;
+	drgn_object_init(&idr_rt, drgn_object_program(idr));
+	drgn_object_init(&idr_base, drgn_object_program(idr));
+
+	err = drgn_object_member(&idr_base, idr, "idr_base");
+	if (!err) {
+		err = drgn_object_read_unsigned(&idr_base, &iter->base);
+		if (err)
+			goto out;
+	} else if (err->code == DRGN_ERROR_LOOKUP) {
+		drgn_error_destroy(err);
+		err = NULL;
+		iter->base = 0;
+	} else {
+		goto out;
+	}
+
+	err = drgn_object_member(&idr_rt, idr, "idr_rt");
+	if (err)
+		goto out;
+	err = drgn_object_address_of(&idr_rt, &idr_rt);
+	if (err)
+		goto out;
+	err = radix_tree_iter_init(&iter->iter, &idr_rt);
+out:
+	drgn_object_deinit(&idr_rt);
+	drgn_object_deinit(&idr_base);
+	return err;
+}
+
+void idr_iter_deinit(struct idr_iter *iter)
+{
+	radix_tree_iter_deinit(&iter->iter);
+}
+
+struct drgn_error *idr_iter_next(struct idr_iter *iter,
+				 struct radix_tree_iter_entry **ret)
+{
+	struct drgn_error *err = radix_tree_iter_next(&iter->iter, ret);
+	if (!err && *ret)
+		(*ret)->index += iter->base;
+	return err;
+}
+
+static struct drgn_error *pid_iter_init_old_kernel(struct drgn_program *prog,
+						   const struct drgn_object *ns,
+						   struct pid_iter *iter)
+{
+	struct drgn_error *err;
+	struct drgn_object pidhash_shift;
+	struct drgn_object ns_level;
+	iter->has_idr = false;
+	iter->index = -1;
+	drgn_object_init(&iter->pid_hash, prog);
+	drgn_object_init(&iter->pos, prog);
+	drgn_object_init(&iter->ns, prog);
+	drgn_object_init(&iter->entry, prog);
+	drgn_object_init(&pidhash_shift, prog);
+	drgn_object_init(&ns_level, prog);
```

Comment:
Similar thing here, these can be just one `tmp` object. (Doing this is really a microoptimization, but if we can trivially save some stack space/time, we might as well.)

Reply:
See [my reply above](https://github.com/osandov/drgn/pull/129#discussion_r770106210)